### ISSUE TITLE
License out of date?

### ISSUE THREAD

**Issue by ConnorLinfoot** [hlandau/acme#191]
Getting the below when trying to request any cert

`{
  "type": "urn:acme:error:malformed",
  "detail": "Provided agreement URL [https://letsencrypt.org/documents/LE-SA-v1.0.1-July-27-2015.pdf] does not match current agreement URL [https://letsencrypt.org/documents/LE-SA-v1.1.1-August-1-2016.pdf]",
  "status": 400
}`

**Comment by zcalusic** [comment]
Same thing here. :(

Tried putting "acme-agreement:https://letsencrypt.org/documents/LE-SA-v1.1.1-August-1-2016.pdf": true in /var/lib/acme/conf.responses, but it didn't help.

Debian package, 0.0.54-1

**Comment by dol** [comment]
Same thing. Tried to add the following two lines to responses file.
```                                                                                                                        
"acme-agreement:https://letsencrypt.org/documents/LE-SA-v1.0.1-July-27-2015.pdf": true                                                                                  
"acme-agreement:https://letsencrypt.org/documents/LE-SA-v1.1.1-August-1-2016.pdf": true 
```

But no help

**Comment by vkijensky** [comment]
I've got the same error:

`{
  "type": "urn:acme:error:malformed",
  "detail": "Provided agreement URL [https://letsencrypt.org/documents/LE-SA-v1.0.1-July-27-2015.pdf] does not match current agreement URL [https://letsencrypt.org/documents/LE-SA-v1.1.1-August-1-2016.pdf]",
  "status": 400
}`

**Comment by chrigl** [comment]
With removing `accounts/acme-v01.api.letsencrypt.org%2fdirectory`, a rerun of reconcile worked.

**Comment by dol** [comment]
@chrigl I'm not sure this is a good advice. This generate a new ACME account and performs a new ToC acceptance. This is why your workaround works.
When using a new ACME account (or deleting the currently used), you lose control over all issued certificates and settings.

**Comment by asquelt** [comment]
:+1: you also loose all your authorizations, so you won't be able to revoke certs if needed. horrible advice.

**Comment by asquelt** [comment]
this is weird...
```
bash-4.3# acmetool quickstart --xlog.severity=debug
20160802111206 [DEBUG] acme.api: request: https://acme-v01.api.letsencrypt.org/directory
 0 / 100 [--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------]   0.00%20160802111206 [DEBUG] acme.api: response: &{200 OK 200 HTTP/1.1 1 1 map[Strict-Transport-Security:[max-age=604800] Cache-Control:[max-age=0, no-cache, no-store] Date:[Tue, 02 Aug 2016 11:12:06 GMT] Connection:[keep-alive] Server:[nginx] Content-Length:[280] X-Frame-Options:[DENY] Expires:[Tue, 02 Aug 2016 11:12:06 GMT] Pragma:[no-cache] Content-Type:[application/json] Boulder-Request-Id:[XXX] Replay-Nonce:[XXX]] 0xc8206bc100 280 [] false map[] 0xc8200fd180 0xc820442d10} <nil>
20160802111206 [DEBUG] acme.api: request: https://acme-v01.api.letsencrypt.org/acme/new-reg
20160802111206 [DEBUG] acme.api: response: &{409 Conflict 409 HTTP/1.1 1 1 map[Replay-Nonce:[XXX] Expires:[Tue, 02 Aug 2016 11:12:06 GMT] Cache-Control:[max-age=0, no-cache, no-store] Date:[Tue, 02 Aug 2016 11:12:06 GMT] Server:[nginx] Location:[https://acme-v01.api.letsencrypt.org/acme/reg/12345] Boulder-Request-Id:[XXX] Boulder-Requester:[12345] Pragma:[no-cache] Content-Type:[application/problem+json] Content-Length:[107]] 0xc820264200 107 [] true map[] 0xc820750000 0xc820442d10} <nil>
20160802111206 [DEBUG] acme.api: request: https://acme-v01.api.letsencrypt.org/acme/reg/12345
20160802111206 [DEBUG] acme.api: response: &{202 Accepted 202 HTTP/1.1 1 1 map[Connection:[keep-alive] Server:[nginx] Content-Length:[617] Cache-Control:[max-age=0, no-cache, no-store] Date:[Tue, 02 Aug 2016 11:12:06 GMT] Replay-Nonce:[XXX] Expires:[Tue, 02 Aug 2016 11:12:06 GMT] Pragma:[no-cache] Content-Type:[application/json] Boulder-Request-Id:[XXX] Boulder-Requester:[12345] Link:[<https://acme-v01.api.letsencrypt.org/acme/new-authz>;rel="next" <https://letsencrypt.org/documents/LE-SA-v1.1.1-August-1-2016.pdf>;rel="terms-of-service"]] 0xc82049c980 617 [] false map[] 0xc82067f340 0xc8203d46e0} <nil>
20160802111206 [DEBUG] acme.api: request: https://acme-v01.api.letsencrypt.org/acme/reg/12345
20160802111207 [DEBUG] acme.api: response: &{400 Bad Request 400 HTTP/1.1 1 1 map[Content-Length:[265] Boulder-Requester:[12345] Pragma:[no-cache] Server:[nginx] Content-Type:[application/problem+json] Boulder-Request-Id:[XXX] Replay-Nonce:[XXX] Expires:[Tue, 02 Aug 2016 11:12:07 GMT] Cache-Control:[max-age=0, no-cache, no-store] Date:[Tue, 02 Aug 2016 11:12:07 GMT]] 0xc820495580 265 [] true map[] 0xc82075ce00 0xc8203d46e0} <nil>
20160802111207 [CRITICAL] acmetool: fatal: couldn't complete registration: HTTP error: 400 Bad Request
map[Expires:[Tue, 02 Aug 2016 11:12:07 GMT] Cache-Control:[max-age=0, no-cache, no-store] Date:[Tue, 02 Aug 2016 11:12:07 GMT] Server:[nginx] Content-Type:[application/problem+json] Boulder-Request-Id:[XXX] Replay-Nonce:[XXX] Content-Length:[265] Boulder-Requester:[12345] Pragma:[no-cache]]
{¶
  "type": "urn:acme:error:malformed",
  "detail": "Provided agreement URL [https://letsencrypt.org/documents/LE-SA-v1.0.1-July-27-2015.pdf] does not match current agreement URL [https://letsencrypt.org/documents/LE-SA-v1.1.1-August-1-2016.pdf]",
  "status": 400
}
```

**Comment by md2k** [comment]
https://github.com/lukas2511/letsencrypt.sh/blob/master/docs/troubleshooting.md#provided-agreement-url-license1-does-not-match-current-agreement-url-license2

Question, acmetool have any options to set that?

**Comment by zcalusic** [comment]
@md2k, we tried, but it didn't work. :(

**Comment by md2k** [comment]
just for info: official certbot working as expected with old license agreement.

**Comment by md2k** [comment]
Looks like problem some where around:   
https://github.com/hlandau/acme/blob/7307a5d7cd8bd308acd0f3e3c0c36483023e4332/acmeapi/api.go#L329-L349    

It return Error in line `338`:
`Registration requires agreement with the following agreement: "https://letsencrypt.org/documents/LE-SA-v1.1.1-August-1-2016.pdf"`

end exit to:
https://github.com/hlandau/acme/blob/ad2cff713d0bf3a1ab03d0e257477877e35489de/solver/register.go#L25

Then as it should execution hit this part of the code https://github.com/hlandau/acme/blob/ad2cff713d0bf3a1ab03d0e257477877e35489de/solver/register.go#L26

Forming response and after `continue` at https://github.com/hlandau/acme/blob/ad2cff713d0bf3a1ab03d0e257477877e35489de/solver/register.go#L58
we start execution of cl.AgreeRegistration at line 24 and we get 400 Bad request with error:
```
{
  "type": "urn:acme:error:malformed",
  "detail": "Provided agreement URL [https://letsencrypt.org/documents/LE-SA-v1.0.1-July-27-2015.pdf] does not match current agreement URL [https://letsencrypt.org/documents/LE-SA-v1.1.1-August-1-2016.pdf]",
  "status": 400
}
```

So basically our problem is here https://github.com/hlandau/acme/blob/ad2cff713d0bf3a1ab03d0e257477877e35489de/solver/register.go#L43-L58

This code set email, but not updates agreement in `reg` object.
```
{
    "resource": "reg",
    "key": {
        "kty": "RSA",
        "n": "KEYDATA",
        "e": "AQAB"
    },
    "contact": [
        "mailto:noc@xxxxxxx.com"
    ],
    "agreement": "https://letsencrypt.org/documents/LE-SA-v1.0.1-July-27-2015.pdf"
}
```
Before trigger AgreeRegistration once again.

**Comment by md2k** [comment]
Also ass addition: 
https://github.com/hlandau/acme/blob/7307a5d7cd8bd308acd0f3e3c0c36483023e4332/acmeapi/api.go#L329-L349

`400 Bad Request` occurred  here https://github.com/hlandau/acme/blob/7307a5d7cd8bd308acd0f3e3c0c36483023e4332/acmeapi/api.go#L330-L333

**Comment by md2k** [comment]
Added between https://github.com/hlandau/acme/blob/ad2cff713d0bf3a1ab03d0e257477877e35489de/solver/register.go#L43-L44
code below solved my issue/
```
reg.AgreementURI = e.URI
```

But i not advice to use it, because not fully familiar with ACME api.

**Comment by asquelt** [comment]
@hlandau can you release it?

**Comment by processwired** [comment]
The first one of my many certificates is going to expire in 27 days... acmetool is a great software, but I wonder what can be done if this issue can't be fixed within this relatively short timeframe?

**Comment by ConnorLinfoot** [comment]
@processwired I'm already in the position where some of my sites SSL has expired, they're not too important but it would be good to get this working as soon as possible.

**Comment by zcalusic** [comment]
I implemented @md2k fix, it seems to work. I confirmed new licence interactively once, and on the next runs it doesn't ask questions anymore. Of course, this required recompilation, and using the compiled binary instead of the debian package provided binary.

**Comment by md2k** [comment]
I can provide build for Linux only, if you want with applied patch.

**Comment by siebenmann** [comment]
Looking at the code, I think the likely bug is in acmeapi/api.go's AgreeRegistration() function. I think it's supposed to submit a registration request with the current data and then if it fails, retry it with the latest agreement URL if you've agreed to it. However the current code inverts a condition; it says 'if the initial request has an error, just return the error' instead of 'if the initial request has no error, we're done'. This is: https://github.com/hlandau/acme/blob/8a25c5288047507147b0c56695357d3569055841/acmeapi/api.go#L331

I changed the 'if err != nil { return err }' section to 'if err == nil { return err }', rebuilt acmetool, and the rebuilt version renewed my certificates without problems. However, I created a response file with the necessary magic to auto-agree and auto-supply my email address, so I may be following a slightly different code path than most people.

If I understand the code right, @md2k's fix avoids this section of the code entirely by overwriting the registration's agreement URL with the latest one you've agreed to right away, instead of leaving it up to AgreeRegistration() to fix this up. The whole area feels a little bit tangled to me, so I'm not sure what I'd call the right fix.

**Comment by farcaller** [comment]
> I changed the 'if err != nil { return err }' section to 'if err == nil { return err }', rebuilt acmetool, and the rebuilt version renewed my certificates without problems

You would want to negate the final `return nil` to `return err` too then.

**Comment by siebenmann** [comment]
Good catch! You're entirely right, and that would have been a rather mysterious failure if (or when) it triggered. Thank you.

**Comment by md2k** [comment]
@siebenmann, nope you got my fix a bit wrong, when it use auto response, acme tool fill only email but agreement kept old, it is old, because ACME return payload of your registration object and agreement you currently using. So i added line which rewrite payload which is assembled few lines earlier with new agreement. it will work for auto-response and for manual response as well.

**Comment by md2k** [comment]
This : https://github.com/hlandau/acme/blob/ad2cff713d0bf3a1ab03d0e257477877e35489de/solver/register.go#L27-L38
is what you will send back to ACME when you pressing 'YES' or with auto-response.
This: https://github.com/hlandau/acme/blob/ad2cff713d0bf3a1ab03d0e257477877e35489de/solver/register.go#L43-L59
Is where is magic happen when `res` not Cacnelled, and response updated with Email, but.. ooops , not with new Agreement. 
So `reg.AgreementURI = e.URI` simply add to `res` this agreement as it should, but wont.   

your change by the way broke the logic of Acmetool, and may cause weird side effects. it good that it works well for you, but acmetool call function you mentioned twice, first when you get response that you use old agreement and should agreed on new one, 2nd time when you agreed and sending back new payload with new agreement. So bug if it exists it is only in part where i se my mini-fix, because if you check more in code above by links, you will see it set new agreement to textual payload but not to res.AgreementURI parameter.

**Comment by md2k** [comment]
My bad your fix will work, but that means you will always automatically accept agreements, way how it was done initially in acmetool, let you chance to disagree with it and stop next actions.
At list this how i see current implementation of those 2 functions.

**Comment by CyBeRoni** [comment]
> but that means you will always automatically accept agreements

Good for automation I say.

**Comment by md2k** [comment]
Just imaging if they write in it that they own your liver and you automatically agreed on it :D

**Comment by zcalusic** [comment]
Nah, blame it on the bug in the program, no liver for them. :)

Jokes aside, it is beautiful to see how we collectively deal with the issue, even though the original developer seems absent for a long time now.

**Comment by petercolberg** [comment]
Could you please resolve hlandau/goutils#1 so that Debian and other distributions that ship acmetool are permitted to distribute the new release that will include this fix?
